File: 1219-path-with-maximum-gold/1219-path-with-maximum-gold.py
class Solution:
    def getMaximumGold(self, grid: List[List[int]]) -> int:
        
        r,c = len(grid), len(grid[0])
        def dfs(i,j,gold):
            nonlocal res
            if i<0 or j<0 or i>=r or j>=c or (i,j) in visited or grid[i][j] == 0:
                res = max(res,gold)
                return 0
            visited.add((i,j))
            dfs(i+1,j,gold+grid[i][j])
            dfs(i-1,j,gold+grid[i][j])
            dfs(i,j+1,gold+grid[i][j])
            dfs(i,j-1,gold+grid[i][j])
            visited.remove((i,j))
            return res
        visited = set()
        res,rest = 0,0
        for i in range(r):
            for j in range(c):
                t = dfs(i,j,0)
                rest = max(rest, t)
        return rest

File: 1219-path-with-maximum-gold/1219-path-with-maximum-gold.cpp
class Solution {
private:
    int dy[4] = {0,0,1,-1}; // For traveling in all 4 direction
    int dx[4] = {1,-1,0,0};
    int maxi = 0; // For maintaining maximum gold
    bool Issafe(vector<vector<int>>& grid, vector<vector<int>> &vis, int r, int c, int m, int n){
        return r>=0 and c>=0 and r<m and c<n and vis[r][c]==0 and grid[r][c]!=0;  // For checking for Safety 
    }
    void dfs(vector<vector<int>>& grid, vector<vector<int>> &vis, int r, int c, int m, int n, int cnt){
        vis[r][c] = 1;
        cnt += grid[r][c];
        for(int k=0; k<4; k++){  // Tra el in all 4  Direction
            int nr = r+dy[k];
            int nc = c+dx[k];
            if(Issafe(grid, vis, nr, nc, m, n)){
                dfs(grid,vis, nr, nc, m, n, cnt);  
            }
            maxi = max(maxi, cnt);
        }
        vis[r][c] = 0; // Backtrack for travel again
    }
public:
    int getMaximumGold(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int>> vis(m, vector<int>(n,0));
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j]!=0){
                    int cnt = 0;
                    dfs(grid,vis, i, j, m, n, cnt);
                }
            }
        }
        return maxi;
    }
};